### PR TITLE
Implementation of the incremental Kolmogorov-Smirnov Statistics

### PR DESCRIPTION
**Pull request by hoanganhngo610**
This implementation features the incremental KS statistics, which is first proposed in the [KDD 2016 paper](https://www.kdd.org/kdd2016/papers/files/rpp0427-dos-reisA.pdf). 

This implementation is based on Treap (or Cartesian Tree), with bulk operation and lazy propagation. This allows the implementation to conduct insertion and removal operations in $O(log N)$ with high probability and calculate the KS test in $O(1)$, a significant improvement compared to the $O(N log N)$ within the non-incremental implementation.

As said, this implementation supports both *insertion* (via `update`) and *deletion* (via `revert`). Moreover, a closely related version of the KS Statistics, the Kuiper test, can also be calculated by setting `statistic` to `kuiper`.

### CODE REVIEW

**Comment by hoanganhngo610** [comment]
@MaxHalford Would you mind having a look through the PR? I think the only remaining problem would be the fact that `ruff` is continuously failing.

File: river/stats/test_ks.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from __future__ import annotations
```
Comment (MaxHalford) [inline river/stats/test_ks.py:1]: Good job on adding a unit test!

File: river/stats/test_ks.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    sliding_b = deque(initial_b)
+
+    for a, b in zip(stream_a, stream_b):
+        sliding_a.popleft()
+        sliding_b.popleft()
+        sliding_a.append(a)
+        sliding_b.append(b)
+        ks_2samp_statistics.append(ks_2samp(sliding_a, sliding_b).statistic)
+
+    assert np.all(np.isclose(np.array(incremental_ks_statistics), np.array(ks_2samp_statistics)))
```
Comment (MaxHalford) [inline river/stats/test_ks.py:43]:
```suggestion
    assert np.allclose(np.array(incremental_ks_statistics), np.array(ks_2samp_statistics))
```

File: river/stats/kolmogorov_smirnov.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+    References
+    ----------
+    [^1]: dos Reis, D.M. et al. (2016) ‘Fast unsupervised online drift detection using incremental Kolmogorov-Smirnov
+    test’, Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge Discovery and Data Mining.
+    doi:10.1145/2939672.2939836.
+    [^2]: C. R. Aragon and R. G. Seidel. Randomized search trees. In FOCS, pages 540–545. IEEE, 1989.
+    [^3]: Kuiper, N. H. (1960). "Tests concerning random points on a circle".
+    Proceedings of the Koninklijke Nederlandse Akademie van Wetenschappen, Series A. 63: 38–47.
+    """
```
Comment (MaxHalford) [inline river/stats/kolmogorov_smirnov.py:221]:
```suggestion
    Proceedings of the Koninklijke Nederlandse Akademie van Wetenschappen, Series A. 63: 38–47.
    
    """
```

File: river/stats/kolmogorov_smirnov.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+            if right_l is not None and right_l.key == key:
+                Treap.sum_all(right, -1 if key[1] == 0 else 1)
+            else:
+                right = Treap.merge(right_l, right)
+
+            self.treap = Treap.merge(left, right)
+
+    def get(self):
+        assert self.statistic in ["ks", "kuiper"]
```
Comment (MaxHalford) [inline river/stats/kolmogorov_smirnov.py:273]: You can remove the assert here. What you should do is add an `else` to handle an unknown string

File: river/stats/kolmogorov_smirnov.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        if self.n_samples == 0:
+            return 0
+
+        if self.statistic == "ks":
+            return max(self.treap.max_value, -self.treap.min_value) / self.n_samples
+        else:
+            return max(self.treap.max_value - self.treap.min_value) / self.n_samples
```
Comment (MaxHalford) [inline river/stats/kolmogorov_smirnov.py:280]:
```suggestion
        elif self.statistic == "kuiper":
            return max(self.treap.max_value - self.treap.min_value) / self.n_samples
        else:
            raise ValueError(f"Unknown statistic {self.statistic}, expected one of: ks, kuiper")
```

**Review by MaxHalford — CHANGES_REQUESTED** [review]
Great job @hoanganhngo610! It's great that you added a unit test, it makes me trust the code is correct.

Can you add an entry to `UNRELEASED.md`?

File: river/stats/test_ks.py (hunk 1 above)
Comment (hoanganhngo610) [reply river/stats/test_ks.py:1]: Thank you :D

File: river/stats/kolmogorov_smirnov.py (hunk 4 above)
Comment (hoanganhngo610) [reply river/stats/kolmogorov_smirnov.py:274]: Totally agree! Suggestion committed.

**Comment by hoanganhngo610** [comment]
@MaxHalford I have successfully incorporated all of your proposed changes to the PR, and also provided a description within the UNRELEASED.md file. Would you mind if I merge the PR into the repository after all tests have passed?

**Pull request merged**